File: C++/permutations/Permutations.cpp
#include <vector>

using namespace std;

void swap(vector<int>& arr, int i, int y) {
    int tmp = arr[i];
    arr[i] = arr[y];
    arr[y] = tmp;
}

void permutate(int idx, vector<int> &arr, vector<vector<int>>& permutations) {
    if (idx == arr.size() - 1) {
        auto new_vector = vector<int>();
        for (int element : arr) {
            new_vector.push_back(element);
        }
        permutations.push_back(new_vector);
    } else {
        for (int y = idx; y < arr.size(); y++) {
            swap(arr, y, idx);
            permutate(idx + 1, arr, permutations);
            swap(arr, idx, y);
        }
    }
}

File: java/src/com/company/Permutations.java
package com.company;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class Permutations {
    public static void main(String[] args) {
        var p = new Permutations();
        List<List<Integer>> permute = p.permute(new int[]{1, 2, 3});
        System.out.println("done");
    }
    public List<List<Integer>> permute(int[] nums) {
        var permutations = new LinkedList<List<Integer>>();
        permuteHelper(0, nums, permutations);
        return permutations;
    }

    public void swap(int[] arr, int i, int y) {
        int tmp = arr[i];
        arr[i] = arr[y];
        arr[y] = tmp;
    }

    public void permuteHelper(int idx, int[] arr, List<List<Integer>> permutations) {
        if (idx == arr.length - 1) {
            var newArr = new ArrayList<Integer>();
            for (var el : arr) {
                newArr.add(el);
            }
            permutations.add(newArr);
        } else {
            for (int y = idx; y < arr.length; y++) {
                swap(arr, idx, y);
                permuteHelper(idx + 1, arr, permutations);
                swap(arr, y, idx);
            }
        }
    }
}
